File: packages/unity-csharp/Runtime/Chain/HederaAdapter.cs
using System;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using System.Collections.Generic;
using UnityEngine.Networking;
using Newtonsoft.Json;

namespace Cinacoin.Chain
{
    /// <summary>
    /// Hedera chain adapter (Hashgraph).
    /// Uses Hedera Mirror Node REST API for balance and transaction queries.
    /// For transactions, use the Hedera JSON-RPC Relay.
    /// </summary>
    public class HederaAdapter : IChainAdapter
    {
        private string _rpcUrl;
        private string _mirrorUrl;

        public string Namespace => "hedera";
        public string ChainReference => "mainnet";
        public string ChainId => "hedera:mainnet";
        public string ChainName => "Hedera Mainnet";

        public HederaAdapter() { }

        public void Init(string rpcUrl)
        {
            _rpcUrl = string.IsNullOrEmpty(rpcUrl) ? "https://mainnet.hashio.io/api" : rpcUrl;
            _mirrorUrl = "https://mainnet-public.mirrornode.hedera.com/api/v1";
        }

        public async Task<BigInteger> GetBalanceAsync(string address)
        {
            // Hedera account IDs are like "0.0.123456"
            var url = $"{_mirrorUrl}/accounts/{address}";
            var response = await CallJsonAsync<HederaAccountResponse>(url);
            var balance = response.Balance?.Balance ?? 0;
            return new BigInteger(balance);
        }

        public async Task<string> GetBalanceFormattedAsync(string address)
        {
            var tinybars = await GetBalanceAsync(address);
            return ((double)tinybars / 1e8).ToString("F8");
        }

        public async Task<long> GetLatestBlockAsync()
        {
            var url = $"{_mirrorUrl}/blocks?order=desc&limit=1";
            var response = await CallJsonAsync<HederaBlockListResponse>(url);
            return response.Blocks?.Count > 0 ? response.Blocks[0].Number : 0;
        }

        public async Task<BigInteger> EstimateFeeAsync(string from, string to, string data = null)
        {
            // Hedera fees are typically very low: ~0.001 HBAR for basic transactions
            return new BigInteger(100_000); // 0.001 HBAR in tinybars
        }

        public async Task<string> SendTransactionAsync(string signedTx)
        {
            // Use Hedera JSON-RPC relay for sending (EVM-compatible)
            var response = await PostJsonAsync<HederaTxResult>(
                _rpcUrl,
                new { jsonrpc = "2.0", method = "eth_sendRawTransaction", @params = new[] { signedTx }, id = 1 });
            return response.Result;
        }

        public async Task<string> GetTransactionAsync(string txHash)
        {
            var url = $"{_mirrorUrl}/transactions/{txHash}";
            return await CallJsonAsync<string>(url);
        }

        public bool IsValidAddress(string address)
        {
            if (string.IsNullOrEmpty(address)) return false;
            // Hedera account ID format: shard.realm.num (e.g. "0.0.123456")
            if (address.StartsWith("0.0.")) return true;
            // Solidity address format: 0x + 40 hex chars
            if (address.StartsWith("0x") && address.Length == 42) return true;
            return false;
        }

        private async Task<T> CallJsonAsync<T>(string url)
        {
            using var request = new UnityWebRequest(url, "GET")
            {
                downloadHandler = new DownloadHandlerBuffer(),
                timeout = 30
            };
            var operation = request.SendWebRequest();
            while (!operation.isDone) await Task.Delay(1);

            if (request.result != UnityWebRequest.Result.Success)
                throw new Exception($"Hedera request failed: {request.error}");

            return JsonConvert.DeserializeObject<T>(request.downloadHandler.text);
        }

        private async Task<T> PostJsonAsync<T>(string url, object body)
        {
            var json = JsonConvert.SerializeObject(body);
            using var request = new UnityWebRequest(url, "POST")
            {
                uploadHandler = new UploadHandlerRaw(Encoding.UTF8.GetBytes(json)),
                downloadHandler = new DownloadHandlerBuffer(),
                timeout = 30
            };
            request.SetRequestHeader("Content-Type", "application/json");

            var operation = request.SendWebRequest();
            while (!operation.isDone) await Task.Delay(1);

            if (request.result != UnityWebRequest.Result.Success)
                throw new Exception($"Hedera POST failed: {request.error}");

            return JsonConvert.DeserializeObject<T>(request.downloadHandler.text);
        }
    }

    internal class HederaAccountResponse
    {
        [JsonProperty("balance")]
        public HederaBalance Balance { get; set; }
        [JsonProperty("account")]
        public string Account { get; set; }
    }

    internal class HederaBalance
    {
        [JsonProperty("balance")]
        public long Balance { get; set; }
        [JsonProperty("timestamp")]
        public string Timestamp { get; set; }
    }

    internal class HederaBlockListResponse
    {
        [JsonProperty("blocks")]
        public List<HederaBlock> Blocks { get; set; }
    }

    internal class HederaBlock
    {
        [JsonProperty("number")]
        public long Number { get; set; }
    }

    internal class HederaTxResult
    {
        [JsonProperty("result")]
        public string Result { get; set; }
    }
}

File: onux/packages/android-kotlin/src/main/kotlin/com/cinacoin/chain/HederaAdapter.kt
/**
 * HederaAdapter — Hedera (HBAR) chain adapter for Android.
 *
 * Provides Hedera-specific operations via Mirror Node API and Consensus Node:
 * - HBAR balance fetching
 * - Transaction broadcasting
 * - Token balance queries
 * - Transaction record retrieval
 *
 * ## Usage
 * ```kotlin
 * val adapter = HederaAdapter("https://mainnet.mirrornode.hedera.com")
 * val balance = adapter.getBalance("0.0.12345")
 * val txId = adapter.sendTransaction("signedTx...")
 * ```
 */
package com.cinacoin.chain


import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*
import java.math.BigDecimal
import java.net.HttpURLConnection
import java.net.URL

// ────────────────────────────────────────────────────────────────────────────
// Hedera Mirror Node Types
// ────────────────────────────────────────────────────────────────────────────

@Serializable
data class HederaAccountResponse(
    val account: String = "",
    val balance: HederaBalance? = null,
    val alias: String? = null,
    val created_timestamp: String = "",
    val key: HederaKey? = null,
    val decline_reward: Boolean = false,
    val ethereum_nonce: Int = 0,
    val evm_address: String? = null,
    val receiver_sig_required: Boolean = false,
    val max_automatic_token_associations: Int = 0,
    memo: String = ""
)

@Serializable
data class HederaBalance(
    val balance: Long = 0,
    val timestamp: String = "",
    val tokens: List<HederaTokenBalance> = emptyList()
)

@Serializable
data class HederaTokenBalance(
    val token_id: String = "",
    val balance: Long = 0
)

@Serializable
data class HederaKey(
    val key: String = "",
    val _type: String = ""
)

@Serializable
data class HederaSendTxResponse(
    val node: String? = null,
    val hash: String = "",
    val transaction: String? = null
)

@Serializable
data class HederaTxRecordResponse(
    val transactions: List<HederaTxRecord> = emptyList(),
    val links: HederaLinks? = null
)

@Serializable
data class HederaTxRecord(
    val consensus_timestamp: String = "",
    val transaction_id: String = "",
    val type: String = "",
    val result: String = "",
    val charged_tx_fee: Long = 0,
    val memo: String = "",
    val transfers: List<HederaTransfer> = emptyList(),
    val token_transfers: List<HederaTokenTransfer> = emptyList()
)

@Serializable
data class HederaTransfer(
    val account: String = "",
    val amount: Long = 0,
    val is_approval: Boolean = false
)

@Serializable
data class HederaTokenTransfer(
    val token_id: String = "",
    val account: String = "",
    val amount: Long = 0,
    val is_approval: Boolean = false
)

@Serializable
data class HederaLinks(
    val next: String? = null
)

@Serializable
data class HederaTokenInfo(
    val token_id: String = "",
    val name: String = "",
    val symbol: String = "",
    val type: String = "",
    val decimals: Int = 0,
    val total_supply: String = "",
    val created_timestamp: String = ""
)

// ────────────────────────────────────────────────────────────────────────────
// Errors
// ────────────────────────────────────────────────────────────────────────────

sealed class HederaError : Exception() {
    data class InvalidAddress(val address: String) : HederaError()
    object NotConnected : HederaError()
    data class InvalidRpcUrl(val url: String) : HederaError()
    data class RpcError(val message: String) : HederaError()
    object InvalidParams : HederaError()
    data class UnsupportedMethod(val method: String) : HederaError()
    data class NotImplemented(val message: String) : HederaError()

    override val message: String
        get() = when (this) {
            is InvalidAddress -> "Invalid Hedera account ID: $address"
            NotConnected -> "Not connected to a Hedera wallet"
            is InvalidRpcUrl -> "Invalid Hedera Mirror Node URL: $url"
            is RpcError -> "Hedera API error: $message"
            InvalidParams -> "Invalid parameters"
            is UnsupportedMethod -> "Unsupported method: $method"
            is NotImplemented -> "Not implemented: $message"
        }
}

// ────────────────────────────────────────────────────────────────────────────
// Hedera Adapter
// ────────────────────────────────────────────────────────────────────────────

class HederaAdapter(
    private var rpcUrl: String = "https://mainnet.mirrornode.hedera.com",
    private val consensusNodeUrl: String = "mainnet-public.hedera.com:50211",
    private val httpClient: HederaHttpClient = DefaultHederaHttpClient()
) : ChainAdapter {

    private var _connectedAddress: String? = null

    override val chainName: String = "hedera"

    override var endpoint: String
        get() = rpcUrl
        set(value) { rpcUrl = value }

    override val isConnected: Boolean
        get() = _connectedAddress != null

    override val connectedAddress: String?
        get() = _connectedAddress

    companion object {
        val MAINNET_MIRROR = "https://mainnet.mirrornode.hedera.com"
        val TESTNET_MIRROR = "https://testnet.mirrornode.hedera.com"
        val PREVIEWNET_MIRROR = "https://previewnet.mirrornode.hedera.com"

        val MAINNET_CONSENSUS = "mainnet-public.hedera.com:50211"
        val TESTNET_CONSENSUS = "testnet.hedera.com:50211"

        /** Validate a Hedera account ID (0.0.XXXX format). */
        fun isValidAddress(address: String): Boolean {
            return address.matches(Regex("^0\\.0\\.\\d+$"))
        }

        /** Convert tinybar to HBAR string. */
        fun tinybarToHbar(tinybar: Long): String {
            val amount = BigDecimal(tinybar)
            return amount.divide(BigDecimal("100000000"), 8, java.math.RoundingMode.HALF_UP)
                .toPlainString()
        }
    }

    // ─── Balance ────────────────────────────────────────────────────────

    /**
     * Get HBAR balance for an account via Mirror Node API.
     * @return Balance in HBAR as decimal string.
     */
    override suspend fun getBalance(accountId: String): String =
        withContext(Dispatchers.IO) {
            require(isValidAddress(accountId)) { "Invalid Hedera account ID: $accountId" }

            val url = "$rpcUrl/api/v1/accounts/$accountId"
            val responseBody = httpClient.get(url)
            val response = Json.decodeFromString<HederaAccountResponse>(responseBody)

            response.balance?.let { tinybarToHbar(it.balance) } ?: "0.00000000"
        }

    // ─── Send Transaction ───────────────────────────────────────────────

    /**
     * Submit a signed transaction to the Hedera Consensus Node.
     * In practice, you'd use the Hedera SDK to build and sign,
     * then submit via gRPC to the consensus node. This method
     * submits via the mirror node relay for HTTP-based flows.
     *
     * @param signedTx Hex-encoded signed transaction bytes.
     * @return Transaction ID.
     */
    override suspend fun sendTransaction(signedTx: String): String =
        withContext(Dispatchers.IO) {
            val body = """{"transaction":"$signedTx"}"""
            val url = "$rpcUrl/api/v1/transactions"
            val responseBody = httpClient.post(url, body)
            val response = Json.decodeFromString<HederaSendTxResponse>(responseBody)
            response.hash.ifEmpty {
                throw HederaError.RpcError("Invalid transaction response")
            }
        }

    // ─── Block Number ──────────────────────────────────────────────────

    /**
     * Get the latest consensus timestamp (block equivalent).
     */
    override suspend fun getLatestBlock(): Long = withContext(Dispatchers.IO) {
        val url = "$rpcUrl/api/v1/blocks?limit=1&order=desc"
        val responseBody = httpClient.get(url)
        val obj = Json.parseToJsonElement(responseBody).jsonObject
        val blocks = obj["blocks"]?.jsonArray
        if (blocks.isNullOrEmpty()) throw HederaError.RpcError("No blocks found")
        blocks.first().jsonObject["timestamp"]?.jsonPrimitive?.content?.toLongOrNull()
            ?: throw HederaError.RpcError("Invalid block response")
    }

    // ─── Fee Estimation ────────────────────────────────────────────────

    /**
     * Estimate fee for a standard HBAR transfer.
     * Hedera has fixed fees: 0.0001 HBAR per transaction.
     */
    override suspend fun estimateFee(): String = "0.0001"

    // ─── Token Balance ──────────────────────────────────────────────────

    /**
     * Get token balances for an account.
     * @return List of token balances.
     */
    suspend fun getTokenBalance(accountId: String): List<HederaTokenBalance> =
        withContext(Dispatchers.IO) {
            require(isValidAddress(accountId)) { "Invalid Hedera account ID: $accountId" }

            val url = "$rpcUrl/api/v1/accounts/$accountId/tokens"
            val responseBody = httpClient.get(url)
            val obj = Json.parseToJsonElement(responseBody).jsonObject
            val arr = obj["tokens"]?.jsonArray ?: JsonArray(emptyList())
            arr.map {
                Json.decodeFromString<HederaTokenBalance>(it.toString())
            }
        }

    // ─── Transaction Record ─────────────────────────────────────────────

    /**
     * Get transaction record by transaction ID or consensus timestamp.
     */
    suspend fun getTransactionRecord(identifier: String): HederaTxRecord =
        withContext(Dispatchers.IO) {
            val url = "$rpcUrl/api/v1/transactions/$identifier"
            val responseBody = httpClient.get(url)
            val response = Json.decodeFromString<HederaTxRecordResponse>(responseBody)

            if (response.transactions.isEmpty()) {
                throw HederaError.RpcError("Transaction not found: $identifier")
            }

            response.transactions.first()
        }

    /**
     * Get transactions for an account.
     */
    suspend fun getAccountTransactions(accountId: String, limit: Int = 10): List<HederaTxRecord> =
        withContext(Dispatchers.IO) {
            val url = "$rpcUrl/api/v1/transactions?account.id=$accountId&limit=$limit"
            val responseBody = httpClient.get(url)
            val response = Json.decodeFromString<HederaTxRecordResponse>(responseBody)
            response.transactions
        }

    /**
     * Get token info by token ID.
     */
    suspend fun getTokenInfo(tokenId: String): HederaTokenInfo =
        withContext(Dispatchers.IO) {
            val url = "$rpcUrl/api/v1/tokens/$tokenId"
            val responseBody = httpClient.get(url)
            Json.decodeFromString<HederaTokenInfo>(responseBody)
        }

    // ─── Connection Management ──────────────────────────────────────────

    override fun setConnectedAddress(address: String) {
        require(isValidAddress(address)) { "Invalid Hedera account ID: $address" }
        _connectedAddress = address
    }

    override fun disconnect() {
        _connectedAddress = null
    }

    // ─── Unified Request ────────────────────────────────────────────────

    suspend fun request(method: String, params: List<Any?> = emptyList()): JsonElement =
        when (method) {
            "hedera_getBalance" -> JsonPrimitive(getBalance(params.firstOrNull() as? String ?: ""))
            "hedera_sendTransaction" -> JsonPrimitive(sendTransaction(params.firstOrNull() as? String ?: ""))
            "hedera_getBlockNumber" -> JsonPrimitive(getLatestBlock())
            "hedera_estimateFee" -> JsonPrimitive(estimateFee())
            else -> throw HederaError.UnsupportedMethod(method)
        }
}

// ────────────────────────────────────────────────────────────────────────────
// HTTP Client Interface
// ────────────────────────────────────────────────────────────────────────────

interface HederaHttpClient {
    suspend fun get(url: String): String
    suspend fun post(url: String, body: String): String
}

class DefaultHederaHttpClient : HederaHttpClient {
    override suspend fun get(url: String): String =
        withContext(Dispatchers.IO) {
            val connection = URL(url).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "GET"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.connectTimeout = 15_000
                connection.readTimeout = 30_000
                if (connection.responseCode != HttpURLConnection.HTTP_OK) {
                    throw HederaError.RpcError("HTTP ${connection.responseCode}")
                }
                connection.inputStream.bufferedReader().use { it.readText() }
            } finally {
                connection.disconnect()
            }
        }

    override suspend fun post(url: String, body: String): String =
        withContext(Dispatchers.IO) {
            val connection = URL(url).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                connection.connectTimeout = 15_000
                connection.readTimeout = 30_000
                connection.outputStream.write(body.toByteArray(Charsets.UTF_8))
                connection.outputStream.flush()
                if (connection.responseCode != HttpURLConnection.HTTP_OK) {
                    throw HederaError.RpcError("HTTP ${connection.responseCode}")
                }
                connection.inputStream.bufferedReader().use { it.readText() }
            } finally {
                connection.disconnect()
            }
        }
}
